File: src/01_run_simpaths.R
#!/usr/bin/env Rscript

library(data.table)
library(withr)

scenario        <- Sys.getenv("SCENARIO")
first_year      <- Sys.getenv("FIRST_YEAR") |> as.numeric()
last_year       <- Sys.getenv("LAST_YEAR") |> as.numeric()
population      <- Sys.getenv("POPULATION") |> as.numeric()
starting_seed   <- Sys.getenv("STARTING_SEED") |> as.numeric()
runs_per_batch  <- Sys.getenv("RUNS_PER_BATCH") |> as.numeric()
batches         <- Sys.getenv("BATCHES") |> as.numeric()
simpaths_path   <- Sys.getenv("SIMPATHS_PATH")

results_root_path <- here::here("data", "simpaths_output")

simpaths_input_path <- file.path(simpaths_path, "input")

timestamp(suffix = paste(" - Started scenario", scenario))

# Delete files that should be recreated during the setup process
file.remove(file.path(simpaths_input_path, "input.mv.db"))

# Store existing output directories before running the simulation, so we can keep track of which ones are new
output_dirs_before <- list.files(file.path(simpaths_path, "output"), full.names = TRUE)

print("Running SimPaths setup")

with_dir(simpaths_path, sys::exec_wait("java", c(
  "-jar", "multirun.jar",
  "-s", format(first_year),
  "-g", "false",
  "-DBSetup"
)))

for (batch in 1:batches) {
  print(paste("Running SimPaths simulation, batch", batch, "of", batches))
  batch_seed <- starting_seed + (batch - 1) * runs_per_batch
  with_dir(simpaths_path, sys::exec_wait("java", c(
    "-jar", "multirun.jar",
    "-r", format(batch_seed),
    "-p", format(population, scientific = FALSE),
    "-n", format(runs_per_batch, scientific = FALSE),
    "-s", format(first_year),
    "-e", format(last_year),
    "-g", "false"
  )))
}

# Identify the new output directories and store their paths for later use
output_dirs_after <- list.files(file.path(simpaths_path, "output"), full.names = TRUE)
new_output_dirs <- setdiff(output_dirs_after, output_dirs_before)
# Don't include "logs" directory (if there is one)
new_output_dirs <- grep(pattern = "logs$", x = new_output_dirs, invert = TRUE, value = TRUE)

results_path <- file.path(results_root_path, scenario)
if (!dir.exists(results_path)) dir.create(results_path, recursive = TRUE)
writeLines(new_output_dirs, file.path(results_path, "output_dirs.txt"))

timestamp(suffix = paste(" - Finished scenario", scenario))

File: src/01_run_simpaths.py
"""
Run the SimPaths microsimulation in batches for two or more scenarios.
"""

import os
import subprocess
import sys
from datetime import datetime
from pathlib import Path


def timestamp(suffix):
    now = datetime.now().strftime("%a %b %d %H:%M:%S %Y")
    print(f"##------ {now} ------##{suffix}", flush=True)


def env_str(name):
    value = os.environ.get(name)
    if not value:
        sys.exit(f"Missing required environment variable: {name}")
    return value


def env_int(name):
    return int(float(env_str(name)))


def list_output_dirs(path):
    if not path.is_dir():
        return []
    return sorted(path.iterdir())


def run_java(args, cwd):
    subprocess.run(["java", "-jar", "multirun.jar", *args], cwd=cwd, check=True)


def main():
    scenario = env_str("SCENARIO")
    first_year = env_int("FIRST_YEAR")
    last_year = env_int("LAST_YEAR")
    population = env_int("POPULATION")
    starting_seed = env_int("STARTING_SEED")
    runs_per_batch = env_int("RUNS_PER_BATCH")
    batches = env_int("BATCHES")

    # Guards agains missplaced SimPaths model
    simpaths_path = Path(env_str("SIMPATHS_PATH"))
    if not simpaths_path.is_dir():
        sys.exit(f"SIMPATHS_PATH is not a directory: {simpaths_path}")
    if not (simpaths_path / "multirun.jar").is_file():
        sys.exit(f"multirun.jar not found in {simpaths_path}")

    simpaths_input_path = simpaths_path / "input"
    simpaths_output_path = simpaths_path / "output"

    results_root_path = Path(__file__).resolve().parents[1] / "data" / "simpaths_output"

    timestamp(f" - Started scenario {scenario}")

    try:
        (simpaths_input_path / "input.mv.db").unlink(missing_ok=True)

        output_dirs_before = set(list_output_dirs(simpaths_output_path))

        print("Running SimPaths setup", flush=True)

        run_java(["-s", str(first_year), "-g", "false", "-DBSetup"], cwd=simpaths_path)

        for batch in range(1, batches + 1):
            print(f"Running SimPaths simulation, batch {batch} of {batches}", flush=True)
            batch_seed = starting_seed + (batch - 1) * runs_per_batch
            run_java([
                "-r", str(batch_seed),
                "-p", str(population),
                "-n", str(runs_per_batch),
                "-s", str(first_year),
                "-e", str(last_year),
                "-g", "false",
            ], cwd=simpaths_path)

        output_dirs_after = list_output_dirs(simpaths_output_path)
        new_output_dirs = [
            path for path in output_dirs_after
            if path not in output_dirs_before and not path.name.endswith("logs")
        ]

        results_path = results_root_path / scenario
        results_path.mkdir(parents=True, exist_ok=True)
        (results_path / "output_dirs.txt").write_text(
            "".join(f"{path}\n" for path in new_output_dirs)
        )

        timestamp(f" - Finished scenario {scenario}")
    finally:
        # Delete heavy files
        for filename in ("input.mv.db", "input.mv.db.lock"):
            (simpaths_input_path / filename).unlink(missing_ok=True)


if __name__ == "__main__":
    main()
